File: Algorithms/src/1039. Minimum Score Triangulation of Polygon/Code/solution.py
from typing import List

class Solution:
    def minScoreTriangulation(self, values: List[int]) -> int:
        n = len(values)
        if n < 3:
            return 0
        # dp[i][j] = min score for sub-polygon from i to j
        dp = [[0] * n for _ in range(n)]
        
        # compute for all lengths starting from 3
        for length in range(3, n + 1):
            for i in range(0, n - length + 1):
                j = i + length - 1
                best = float('inf')
                for k in range(i + 1, j):
                    cost = dp[i][k] + dp[k][j] + values[i] * values[k] * values[j]
                    if cost < best:
                        best = cost
                dp[i][j] = best
        return dp[0][n-1]

File: Algorithms/src/1039. Minimum Score Triangulation of Polygon/Code/solution.java
class Solution {
    public int minScoreTriangulation(int[] values) {
        int n = values.length;
        int[][] dp = new int[n][n]; // default 0
        
        for (int len = 3; len <= n; len++) {
            for (int i = 0; i + len - 1 < n; i++) {
                int j = i + len - 1;
                int best = Integer.MAX_VALUE;
                for (int k = i + 1; k < j; k++) {
                    int cost = dp[i][k] + dp[k][j] + values[i] * values[k] * values[j];
                    if (cost < best) best = cost;
                }
                dp[i][j] = best;
            }
        }
        return n == 0 ? 0 : dp[0][n-1];
    }
}
